### PR TITLE
Aircraft Generator

### PR DESCRIPTION
**Pull request by german-e-mas**
In this PR I include the random Aircraft generator.
The generation task is fired between 1 and 2 seconds, creating a random type of Aircraft in a random position near the edges of the Map (it wouldn't make sense if they appeared inside, out of nowhere). Their direction is set according to this as well.

I also made some changes in the Map class. It had a lot of static methods. I now instantiate the class and set it's boundaries in the constructor. The method to check if an Aircraft is inside the map is public, but the one to check for `Position` is private. I think it's more important to ask if a certain object is inside the map and have it's position be handled internally. Or having a single public method to check for position, and pass the object's position to it?

ptal @pablisho

### CODE REVIEW

File: app/src/main/java/mas/german/landingplanes/AircraftGenerator.java [hunk 1]
```diff
@@ -0,0 +1,133 @@
...
+    }
+
+    /**
+     * Sets the Listener in the classes that require randomly generated aircrafts.
+     */
+    public void setOnAircraftGeneratedListener(OnAircraftGenerated listener) {
+        mOnAircraftGeneratedListener = listener;
+    }
+
+    AircraftGenerator(Map currentMap) {
```
Comment (pablisho) [inline app/src/main/java/mas/german/landingplanes/AircraftGenerator.java]: I would pass the Game here. At least for me it looks better, then inside the method you can ask for the map and store it

File: app/src/main/java/mas/german/landingplanes/AircraftGenerator.java [hunk 2]
```diff
@@ -0,0 +1,133 @@
...
+    private Aircraft generateRandomAircraft() {
+        Aircraft randomAircraft;
+        Position randomPosition;
+
+        // First calculate the position, as the range of directions also depends on it.
+        // The aircraft starts from a side of the Map, and it's direction is set accordingly.
+        double x = 0;
+        double y = 0;
+        double angle = 0;
+        switch (3) {
```
Comment (pablisho) [inline app/src/main/java/mas/german/landingplanes/AircraftGenerator.java]: switch 3?

File: app/src/main/java/mas/german/landingplanes/AircraftGenerator.java [hunk 3]
```diff
@@ -0,0 +1,133 @@
...
+            case 1:
+                randomAircraft = new LightPlane(mRandom.nextInt(5) + 10, angle, randomPosition);
+                break;
+            case 2:
+                randomAircraft = new Helicopter(mRandom.nextInt(5) + 5, angle, randomPosition);
+                break;
+            default:
+                randomAircraft = new LargePlane(0, angle, randomPosition);
+        }
+        return randomAircraft;
```
Comment (pablisho) [inline app/src/main/java/mas/german/landingplanes/AircraftGenerator.java]: What about velocity? All the planes have the same?

File: app/src/main/java/mas/german/landingplanes/AircraftGenerator.java [hunk 4]
```diff
@@ -0,0 +1,133 @@
...
+        mRandom = new Random(System.currentTimeMillis());
+        mMap = currentMap;
+        // Start a new executor thread.
+        mExecutor = Executors.newScheduledThreadPool(1);
+    }
+
+    /**
+     * Schedules the first task. The task re-schedules itself after some time defined at random.
+     */
+    public void begin() {
```
Comment (pablisho) [inline app/src/main/java/mas/german/landingplanes/AircraftGenerator.java]: it should have a stop too

File: app/src/main/java/mas/german/landingplanes/Game.java [hunk 5]
```diff
@@ -108,4 +115,12 @@ private void gameOver() {
         // Cancel the Update Task.
         mUpdateTask.cancel(true);
     }
+
+    @Override
+    public void onAircraftGenerated(Aircraft generatedAircraft) {
+        // Synchronize the Aircraft list to prevent access during the operation.
+        synchronized (mAircrafts) {
```
Comment (pablisho) [inline app/src/main/java/mas/german/landingplanes/Game.java]: This is ok, but you should also synchronize when accessing this list for reading. Just having one synchronize doesn't prevent the access, unless the other one also has a synchronize block on the same field

File: app/src/main/java/mas/german/landingplanes/Map.java [hunk 6]
```diff
@@ -12,36 +12,57 @@
...
+    private final double mBoundaryRight;
+    private final double mBoundaryBottom;
 
     /**
      * Get the unique instance of the Map class.
      *
      * @return Map instance.
      */
-    public static Map getInstance() {
+    public static Map getInstance(double left, double top, double right, double bottom) {
```
Comment (pablisho) [inline app/src/main/java/mas/german/landingplanes/Map.java]: This should definitely stop being a singleton

File: app/src/main/java/mas/german/landingplanes/AircraftGenerator.java (hunk 2 above)
Comment (german-e-mas) [inline app/src/main/java/mas/german/landingplanes/AircraftGenerator.java]: I missed this correction. I was testing with fixed numbers, to see different creations in one particular side. It should be `mRandom.nextInt(4)`

File: app/src/main/java/mas/german/landingplanes/AircraftGenerator.java (hunk 3 above)
Comment (german-e-mas) [inline app/src/main/java/mas/german/landingplanes/AircraftGenerator.java]: The speed is the first parameter on their creator. It's a random number between a minimum and a maximum, depending on the type. Should these min/max values be a static parameter of each class?

**Comment by german-e-mas** [comment]
I addressed the comments.
- The list `mAircrafts` is synchronized every time something needs access to it.
- The default random Aircraft is `null`. The following switch statement will modify it.
- Map class is no longer singleton.
- Renamed the static final radius of each aircraft. There was no need to state the type in it's name.
- Included minimum and maximum speed values for each type of aircraft as static final integers.
- Included a `stop()` method for the Aircraft generator class, which shuts it down. The `begin()` method can restart the executor thread if it's called again after a shut down.
- Game's `gameOver()` method stops the generation of aircrafts with the previously mentioned method.

ptal @pablisho

File: app/src/main/java/mas/german/landingplanes/aircrafts/Helicopter.java [hunk 7]
```diff
@@ -8,11 +8,14 @@
  * It can only land on helipads.
  */
 public class Helicopter extends Aircraft {
+    public static final int MAX_SPEED = 5;
+    public static final int MIN_SPEED = 5;
```
Comment (pablisho) [inline app/src/main/java/mas/german/landingplanes/aircrafts/Helicopter.java]: Always the same?

File: app/src/main/java/mas/german/landingplanes/aircrafts/LargePlane.java [hunk 8]
```diff
@@ -8,11 +8,14 @@
  * It can only land on long runways.
  */
 public class LargePlane extends Aircraft {
+    public static final int MAX_SPEED = 10;
+    public static final int MIN_SPEED = 10;
```
Comment (pablisho) [inline app/src/main/java/mas/german/landingplanes/aircrafts/LargePlane.java]: same

File: app/src/main/java/mas/german/landingplanes/Map.java [hunk 9]
```diff
@@ -3,45 +3,52 @@
...
-     * Get the unique instance of the Map class.
-     *
-     * @return Map instance.
-     */
-    public static Map getInstance() {
-        if (sInstance == null) {
-            sInstance = new Map();
-        }
-        return sInstance;
+    Map(double left, double top, double right, double bottom) {
```
Comment (pablisho) [inline app/src/main/java/mas/german/landingplanes/Map.java]: Are you sure this names are ok? I mean, do you remember how we decided the coordinate system to be? Y up or down? if Y is down then up boundary should be 0 and down boundary should be 100

**Comment by german-e-mas** [comment]
I addressed the comments about the speed and commented where the Map's origin is. Thanks for the review @pablisho , I'll merge now.

**Pull request merged**